src/BlockDiagonals.jl

```diff
@@ -0,0 +1,338 @@
+        end
+        row += nrows
+    end
+    return A
+end
+
+Base.:+(M::UniformScaling, B::BlockDiagonal) = B + M
+function Base.:+(B::BlockDiagonal, M::UniformScaling)
+    return BlockDiagonal([block + M for block in blocks(B)])
+end
+
+## Multiplication
+Base.:*(n::Real, B::BlockDiagonal) = B * n
+Base.:*(B::BlockDiagonal, n::Real) = BlockDiagonal(n .* blocks(B))
+
+# TODO make type stable, maybe via Broadcasting?
+function Base.:*(B1::BlockDiagonal, B2::BlockDiagonal)
+    if size(B1) == size(B2) && size.(blocks(B1)) == size.(blocks(B2))
+        return BlockDiagonal(blocks(B1) .* blocks(B2))
+    else
+        return Matrix(B1) * Matrix(B2)
+    end
+end
+
+function Base.:*(B::BlockDiagonal, M::AbstractMatrix)::Matrix
+    if size(B, 2) !== size(M, 1)
+        throw(DimensionMismatch("A has dimensions $(size(B)) but B has dimensions $(size(M))"))
+    end
+    st = 1  # start
+    ed = 0  # end
+    d = []
+    for block in blocks(B)
+        ed += size(block, 2)
+        push!(d, block * M[st:ed, :])
+        st = ed + 1
+    end
+    return reduce(vcat, d)
+end
+
+function Base.:*(M::AbstractMatrix, B::BlockDiagonal)
+    if size(M, 2) !== size(B, 1)
+        throw(DimensionMismatch("A has dimensions $(size(M)) but B has dimensions $(size(B))"))
+    end
+    st = 1  # start
+    ed = 0  # end
+    d = []
+    for block in blocks(B)
+        ed += size(block, 1)
+        push!(d, M[:, st:ed] * block)
+        st = ed + 1
+    end
+    return reduce(hcat, d)::Matrix
+end
+
+# Diagonal
+function Base.:*(B::BlockDiagonal, M::Diagonal)::BlockDiagonal
+    if size(B, 2) !== size(M, 1)
+        throw(DimensionMismatch("A has dimensions $(size(B)) but B has dimensions $(size(M))"))
+    end
+    A = copy(B)
+    d = diag(M)
+    col = 1
+    for (p, block) in enumerate(blocks(B))
+        ncols = size(block, 2)
+        cols = col:(col + ncols-1)
+        for (j, c) in enumerate(cols)
+            A[Block(p)][:, j] *= d[c]
+        end
+        col += ncols
+    end
+    return A
+end
+
+function Base.:*(M::Diagonal, B::BlockDiagonal)::BlockDiagonal
+    if size(M, 2) !== size(B, 1)
+        throw(DimensionMismatch("A has dimensions $(size(M)) but B has dimensions $(size(B))"))
+    end
+    A = copy(B)
+    d = diag(M)
+    row = 1
+    for (p, block) in enumerate(blocks(B))
+        nrows = size(block, 1)
+        rows = row:(row + nrows-1)
+        for (i, r) in enumerate(rows)
+            A[Block(p)][i, :] *= d[r]
+        end
+        row += nrows
+    end
+    return A
+end
+
+## Division
+Base.:/(B::BlockDiagonal, n::Real) = BlockDiagonal(blocks(B) ./ n)
+
+## LinearAlgebra
+for f in (:adjoint, :eigvecs, :inv, :transpose)
+    @eval LinearAlgebra.$f(B::BlockDiagonal) = BlockDiagonal(map($f, blocks(B)))
+end
+
+LinearAlgebra.diag(B::BlockDiagonal) = mapreduce(diag, vcat, blocks(B))
+LinearAlgebra.det(B::BlockDiagonal) = prod(det, blocks(B))
+LinearAlgebra.logdet(B::BlockDiagonal) = sum(logdet, blocks(B))
```

Comment:
How hard is it to define `lu!`?
If you do that then you get `det` and `logdet` for free

Reply:
I don't know - we can have `lu(B::BlockDiagonal = lu(Matrix(B))` but not sure how implement `lu!` (there'e probably a nice way tbh, but i'd need to look into it) since the definitions of `*det` are so nice and easy i think it's fine for now?